### PR TITLE
Avoid defining methods inside of methods.

### LINKED ISSUE #437: Add a note about nested function declaration
Unfortunately sometimes I see something like this
```ruby
def foo(x)
  def bar(y)
    do_work y
  end
  bar(x)
end
```
in the wild and it's author doesn't know that it isn't nested function declaration, but full-blown function declaration (and re-declaration on each foo() call).

I think that it won't hurt if there will be mention of this with suggestion to use lambda/Proc or "non-nested" function declaration.

### PR DESCRIPTION
**Pull request by strand**
fixes #437 

Instead use lambdas to define processes at a local scope.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -402,6 +402,31 @@ Translations of the guide are available in the following languages:
   end
   ```
 
+
+* <a name="method-definition-inside-methods"></a>
+    Avoid defining methods inside of methods, as this dynamically defines the "internal" method at the same scope as the "external" method. Instead use a lambda to define a process local to the scope of the method.
```
Comment (bbatsov) [inline README.md]: internal -> nested
external -> outer
process?

File: README.md [hunk 2]
```diff
@@ -402,6 +402,31 @@ Translations of the guide are available in the following languages:
   end
   ```
 
+
+* <a name="method-definition-inside-methods"></a>
```
Comment (bbatsov) [inline README.md]: nested-method-definition perhaps?

File: README.md [hunk 3]
```diff
@@ -402,6 +402,31 @@ Translations of the guide are available in the following languages:
...
+    Avoid defining methods inside of methods, as this dynamically defines the "internal" method at the same scope as the "external" method. Instead use a lambda to define a process local to the scope of the method.
+  <sup>[[link](#method-definition-inside-methods)]</sup>
+
+    ```Ruby
+    # good
+    def some_method(argument)
+      internal_process = lambda do |foobar|
```
Comment (bbatsov) [inline README.md]: process doesn't sound right to me

**Comment by bbatsov** [comment]
Using code examples within a class might make explaining this even easier - http://stackoverflow.com/a/4865161/291550

**Comment by ojab** [comment]
I'm thinking about something like [this](https://github.com/ojab/ruby-style-guide/commit/018091e134ff336488e23ade0c450c7f855f71d1).
IMHO guide should show the actual scope of inner method definition.

No sure if proc should be mentioned though, since it's semantic differs from method/lambda.

**Comment by bbatsov** [comment]
@ojab Overall your suggestion looks good. I wouldn't mention the proc.

**Comment by ojab** [comment]
Hopefully fixed. Should I make PR or my changes will be incorporated into this one? I'm ok with both options.

**Comment by strand** [comment]
Hey @ojab, Just saw all the feedback. I'll integrate it and amend the commit. I'll avoid the word process, rename the anchor, use a class, and use nested and outer instead of internal and external.

**Comment by strand** [comment]
Oh, @ojab I just saw that you've written a commit on this already. Submit a PR and I will close mine.

**Comment by ojab** [comment]
ok, made PR https://github.com/bbatsov/ruby-style-guide/pull/439

**Pull request closed without merging**